From robert.jarzmik@free.fr  Tue Dec 22 11:30:59 2009
From: Enrico Scholz <enrico.scholz@sigma-chemnitz.de>
Date: Fri, 04 Dec 2009 17:50:08 +0100
Subject: USB: pxa27x_udc: avoid compiler warnings and misbehavior on buggy hardware
To: dbrownell@users.sourceforge.net, gregkh@suse.de, Enrico Scholz <enrico.scholz@sigma-chemnitz.de>
Cc: linux-usb@vger.kernel.org, eric.y.miao@gmail.com, linux@arm.linux.org.uk, Jonathan Cameron <jic23@cam.ac.uk>
Message-ID: <87einaiszj.fsf@free.fr>


From: Enrico Scholz <enrico.scholz@sigma-chemnitz.de>
Date: Sun, 11 Oct 2009 11:52:48 +0200

hardware reports wrong interrupt.  Although such a situation should not
happen, the compiler complains about this access.

This patch adds a sanity check and generates warning to detect such
issues.

Signed-off-by: Enrico Scholz <enrico.scholz@sigma-chemnitz.de>
Acked-by: Robert Jarzmik <robert.jarzmik@free.fr>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/pxa27x_udc.c |   19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

--- a/drivers/usb/gadget/pxa27x_udc.c
+++ b/drivers/usb/gadget/pxa27x_udc.c
@@ -2218,9 +2218,13 @@ static void irq_handle_data(int irq, str
 			continue;
 
 		udc_writel(udc, UDCISR0, UDCISR_INT(i, UDCISR_INT_MASK));
-		ep = &udc->pxa_ep[i];
-		ep->stats.irqs++;
-		handle_ep(ep);
+
+		WARN_ON(i >= ARRAY_SIZE(udc->pxa_ep));
+		if (i < ARRAY_SIZE(udc->pxa_ep)) {
+			ep = &udc->pxa_ep[i];
+			ep->stats.irqs++;
+			handle_ep(ep);
+		}
 	}
 
 	for (i = 16; udcisr1 != 0 && i < 24; udcisr1 >>= 2, i++) {
@@ -2228,9 +2232,12 @@ static void irq_handle_data(int irq, str
 		if (!(udcisr1 & UDCISR_INT_MASK))
 			continue;
 
-		ep = &udc->pxa_ep[i];
-		ep->stats.irqs++;
-		handle_ep(ep);
+		WARN_ON(i >= ARRAY_SIZE(udc->pxa_ep));
+		if (i < ARRAY_SIZE(udc->pxa_ep)) {
+			ep = &udc->pxa_ep[i];
+			ep->stats.irqs++;
+			handle_ep(ep);
+		}
 	}
 
 }
